Running rust-toolchain twice in same workflow can result in unstable errors

Example workflow where we use the nightly for formatting, but stable for linting with Clippy:

```
    steps:
      - uses: actions/checkout@v3
      - uses: dtolnay/rust-toolchain@nightly
        with:
          components: rustfmt
      - run: cargo +nightly fmt --all -- --check
      - uses: dtolnay/rust-toolchain@stable
        with:
          components: clippy
      - uses: Swatinem/rust-cache@v2
      - run: cargo clippy --all-targets --all-features --locked -- -D 
```

This results in the following error when trying to run the `clippy` command:

```
  env:
    CARGO_INCREMENTAL: 0
    CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  
error: failed to get `axum` as a dependency of package `CRATE v0.1.0 (/runner/_work/PATH)`

Caused by:
  failed to load source for dependency `axum`

Caused by:
  Unable to update registry `crates-io`

Caused by:
  usage of sparse registries requires `-Z sparse-registry`
Error: Process completed with exit code 101.
```

--- a/action.yml
+++ b/action.yml
@@ -100,9 +100,12 @@ runs:
     - run: |
         : enable Cargo sparse registry
         # except on 1.66 and 1.67, on which it is unstable
-        if [ -z "${CARGO_REGISTRIES_CRATES_IO_PROTOCOL+set}" ]; then
+        if [ -z "${CARGO_REGISTRIES_CRATES_IO_PROTOCOL+set}" -o -f "${{runner.temp}}"/.implicit_cargo_registries_crates_io_protocol ]; then
+          touch "${{runner.temp}}"/.implicit_cargo_registries_crates_io_protocol || true
           if rustc +${{steps.parse.outputs.toolchain}} --version --verbose | (! grep -q '^release: 1\.6[67]\.'); then
             echo CARGO_REGISTRIES_CRATES_IO_PROTOCOL=sparse >> $GITHUB_ENV
+          else
+            echo CARGO_REGISTRIES_CRATES_IO_PROTOCOL=git >> $GITHUB_ENV
           fi
         fi
       shell: bash

